skimage/feature/sift.py

```diff
@@ -0,0 +1,708 @@
+
+        Examples
+        --------
+        >>> from skimage.feature import SIFT, match_descriptors
+        >>> from skimage.data import camera
+        >>> from skimage.transform import rotate
+        >>> img1 = camera()
+        >>> img2 = rotate(camera(), 90)
+        >>> detector_extractor1 = SIFT()
+        >>> detector_extractor2 = SIFT()
+        >>> detector_extractor1.detect_and_extract(img1)
+        >>> detector_extractor2.detect_and_extract(img2)
+        >>> matches = match_descriptors(detector_extractor1.descriptors,
+        ...                             detector_extractor2.descriptors,
+        ...                             max_ratio=0.6)
+        >>> matches[10:15]
+        array([[ 11,  11],
+               [ 12, 568],
+               [ 13,  13],
+               [ 14, 569],
+               [ 15,  15]])
+        >>> detector_extractor1.keypoints[matches[10:15, 0]]
+        array([[170, 241],
+               [341, 287],
+               [234,  13],
+               [232, 378],
+               [206, 307]])
+        >>> detector_extractor2.keypoints[matches[10:15, 1]]
+        array([[271, 170],
+               [383,  95],
+               [499, 234],
+               [191, 260],
+               [205, 206]])
+
+        """
+
+    def __init__(self, upsampling=2, n_octaves=8, n_scales=3, sigma_min=1.6,
+                 sigma_in=0.5,
+                 c_dog=0.04 / 3, c_edge=10, n_bins=36, lambda_ori=1.5,
+                 c_max=0.8, lambda_descr=6, n_hist=4, n_ori=8):
+        if upsampling in [1, 2, 4]:
+            self.upsampling = upsampling
+        else:
+            raise ValueError("upsampling must be 1, 2 or 4")
+        self.n_octaves = n_octaves
+        self.n_scales = n_scales
+        self.sigma_min = sigma_min / upsampling
+        self.sigma_in = sigma_in
+        self.c_dog = (2 ** (1 / n_scales) - 1) / (2 ** (1 / 3) - 1) * c_dog
+        self.c_edge = c_edge
+        self.n_bins = n_bins
+        self.lambda_ori = lambda_ori
+        self.c_max = c_max
+        self.lambda_descr = lambda_descr
+        self.n_hist = n_hist
+        self.n_ori = n_ori
+
+        self.delta_min = 1 / upsampling
+        self.deltas = (self.delta_min
+                       * np.power(2, np.arange(self.n_octaves - 1)))
+        self.scalespace_sigmas = None
+        self.keypoints = None
+        self.positions = None
+        self.sigmas = None
+        self.scales = None
+        self.orientations = None
+        self.octaves = None
+        self.descriptors = None
+
+    def _number_of_octaves(self, n, image_shape):
+        sMin = 12  # minimum size of last octave
+        s0 = np.min(image_shape)
+        return int(
+            np.min((n, (np.log(s0 / sMin) / np.log(2)) + self.upsampling)))
+
+    def _create_scalespace(self, image):
+        """Source: "Anatomy of the SIFT Method" Alg. 1
+        Construction of the scalespace by gradually blurring (scales) and
+        downscaling (octaves) the image.
```

Comment:
You don't have to address this comment immediately, but longer term we should think about how to refactor the scale space code in `skimage/feature/blob.py` so it could be reused here as well. I think we will need first the following fix there
https://github.com/scikit-image/scikit-image/pull/4482
and then I think there was another PR that had started that refactoring. I will link it here when I find it.

Reply:
It was this issue and corresponding PR. It is pretty old, though, so has become a bit out of sync with recent changes:
https://github.com/scikit-image/scikit-image/issues/2315
https://github.com/scikit-image/scikit-image/pull/2386